src/lib/device/journal_vdev.cpp

```diff
@@ -34,75 +34,211 @@
+        m_journal_descriptors.emplace(logdev_id, journal_desc);
+
+        // Traverse the list starting from the head and add those chunks
+        // in order to the journal descriptor. next_chunk is stored in private_data.
+        // Last chunk will have next_chunk as 0.
+        auto chunk_num = head.chunk_num;
+        while (chunk_num != 0) {
+            auto& c = chunk_map[chunk_num];
+            RELEASE_ASSERT(c, "Invalid chunk found logdev {} chunk num {}", logdev_id, chunk_num);
+            journal_desc->m_journal_chunks.push_back(c);
+            visited_chunks.insert(chunk_num);
+
+            // Increase the right window and total size.
+            journal_desc->m_total_size += c->size();
+            journal_desc->m_end_offset += c->size();
+
+            auto data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(c->user_private()));
+            chunk_num = data->next_chunk;
+        }
+    }
 
-    const off_t ds_off = data_start_offset();
-    const off_t end_offset = tail_offset();
+    // Chunks which are not in visited set are orphans and needs to be cleaned up.
+    // Remove chunk will affect the m_all_chunks so keep a separate list.
+    std::vector< shared< Chunk > > orphan_chunks;
+    for (auto& chunk : m_all_chunks) {
+        if (!visited_chunks.count(chunk->chunk_id())) { orphan_chunks.push_back(chunk); }
+    }
 
-    auto const [chunk, offset_in_chunk] = offset_to_chunk(end_offset);
+    for (auto& chunk : orphan_chunks) {
+        auto* data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
+        auto chunk_id = chunk->chunk_id();
+        auto logdev_id = data->logdev_id;
+        auto next_chunk = data->next_chunk;
 
-#ifndef NDEBUG
-    if (end_offset < ds_off) { HS_DBG_ASSERT_EQ(size() - used_size(), static_cast< uint64_t >(ds_off - end_offset)); }
-#endif
-    // works for both "end_offset >= ds_off" and "end_offset < ds_off";
-    if (offset_in_chunk + sz <= chunk->size()) {
-        // not acrossing boundary, nothing to do;
-    } else if ((used_size() + (chunk->size() - offset_in_chunk) + sz) <= size()) {
-        // across chunk boundary, still enough space;
+        // Clear the private chunk data.
+        *data = JournalChunkPrivate{};
+        update_chunk_private(chunk, data);
+
+        // Try add to chunk_pool, if pool is full we remove the chunk.
+        if (m_chunk_pool->size() < m_chunk_pool->capacity()) {
+            m_chunk_pool->enqueue(chunk);
+            LOGWARN("Reuse to pool orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+        } else {
+            LOGWARN("Removing orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+            m_dmgr.remove_chunk(chunk);
+        }
+    }
 
-        // Update the overhead to total write size;
-        m_write_sz_in_total.fetch_add(chunk->size() - offset_in_chunk, std::memory_order_relaxed);
```

Comment:
this line should be kept, no?

Reply:
This is now done in append_chunk() function line 183